### PR TITLE
feature: support large pipeline

### PR DESCRIPTION
**Pull request by navaj0**
*Issue #, if available:*

*Description of changes:*

*Testing done:*

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I certify that the changes I am introducing will be backword compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [x] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [x] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by staubhp — APPROVED** [review]

File: src/sagemaker/workflow/pipeline.py [hunk 1]
```diff
@@ -102,37 +105,62 @@ def create(
...
             PipelineName=self.name,
-            PipelineDefinition=self.definition(),
             RoleArn=role_arn,
         )
+
+        # If pipeline definition is large, upload to S3 bucket and
+        # provide PipelineDefinitionS3Location to request instead.
+        if len(pipeline_definition.encode("utf-8")) < 1024 * 100:
+            kwargs["PipelineDefinition"] = self.definition()
+        else:
```
Comment (mufaddal-rohawala) [inline src/sagemaker/workflow/pipeline.py:148]: is it possible to add test for this else block, to verify s3 sourced pipeline definition works?

**Review by mufaddal-rohawala — CHANGES_REQUESTED** [review]

File: src/sagemaker/workflow/pipeline.py (hunk 1 above)
Comment (navaj0) [reply src/sagemaker/workflow/pipeline.py:148]: Yes. Added more unit tests to cover this branch. Integration tests are already there.

File: src/sagemaker/workflow/pipeline.py [hunk 2]
```diff
@@ -102,37 +105,62 @@ def create(
...
         kwargs = dict(
             PipelineName=self.name,
-            PipelineDefinition=self.definition(),
             RoleArn=role_arn,
         )
+
+        # If pipeline definition is large, upload to S3 bucket and
+        # provide PipelineDefinitionS3Location to request instead.
+        if len(pipeline_definition.encode("utf-8")) < 1024 * 100:
+            kwargs["PipelineDefinition"] = self.definition()
```
[the remark this answers is not in the document]
Comment (navaj0) [reply src/sagemaker/workflow/pipeline.py]: Good catch. Updated it.

File: src/sagemaker/workflow/pipeline.py (hunk 1 above)
Comment (navaj0) [reply src/sagemaker/workflow/pipeline.py:148]: Yes. Added more unit tests to cover the new branch.

File: src/sagemaker/workflow/pipeline.py [hunk 3]
```diff
@@ -102,37 +105,62 @@ def create(
...
+        kwargs = self._create_args(role_arn, description, parallelism_config)
         update_args(
             kwargs,
             Tags=tags,
         )
         return self.sagemaker_session.sagemaker_client.create_pipeline(**kwargs)
 
-    def _create_args(self, role_arn: str, description: str):
+    def _create_args(
+        self, role_arn: str, description: str, parallelism_config: ParallelismConfiguration
```
Comment (jeniyat) [inline src/sagemaker/workflow/pipeline.py:124]: Is it possible to add default values to these parameters?

File: tests/integ/test_workflow.py [hunk 4]
```diff
@@ -2757,3 +2757,99 @@ def cleanup_feature_group(feature_group: FeatureGroup):
...
+    sagemaker_session, role, pipeline_name, region_name
+):
+    instance_count = ParameterInteger(name="InstanceCount", default_value=2)
+
+    outputParam = CallbackOutput(output_name="output", output_type=CallbackOutputTypeEnum.String)
+
+    callback_steps = [
+        CallbackStep(
+            name=f"callback-step{count}",
+            sqs_queue_url="https://sqs.us-east-2.amazonaws.com/123456789012/MyQueue",
```
Comment (jeniyat) [inline tests/integ/test_workflow.py:2817]: Will it be an accessible url even if the user is from a different region (i.e., us-west-2)?

**Review by jeniyat — CHANGES_REQUESTED** [review]

File: tests/integ/test_workflow.py (hunk 4 above)
Comment (navaj0) [reply tests/integ/test_workflow.py:2817]: It is not an accessible url. It is actually just a well formatted dummy URL embedded in the pipeline definition to verify the pipeline creation.

File: src/sagemaker/workflow/pipeline.py (hunk 3 above)
Comment (navaj0) [reply src/sagemaker/workflow/pipeline.py:124]: Actually no. All of them must be explicitly set by the users.

**Review by jeniyat — APPROVED** [review]

**Review by mufaddal-rohawala — APPROVED** [review]

**Pull request merged**